p2p/net/swarm/swarm_listen.go

```diff
@@ -79,24 +97,27 @@ func (s *Swarm) AddListenAddr(a ma.Multiaddr) error {
 	go func() {
 		defer func() {
 			list.Close()
+
 			s.listeners.Lock()
+			defer s.listeners.Unlock()
+
 			delete(s.listeners.m, list)
 			s.listeners.cacheEOL = time.Time{}
-			s.listeners.Unlock()
 
 			// signal to our notifiees on listen close.
 			s.notifyAll(func(n network.Notifiee) {
 				n.ListenClose(s, maddr)
 			})
 			s.refs.Done()
+
+			// log if no more listener and the swarm is still running.
+			if len(s.listeners.m) == 0 && s.ctx.Err() == nil {
+				log.Error(ErrSwarmListenerAcceptError)
+			}
 		}()
 		for {
 			c, err := list.Accept()
 			if err != nil {
-				if s.ctx.Err() == nil {
```

Comment:
This is here to tell the user that a listener failed for some reason, we need to keep it here.

It sounds like we may need a way to mark listeners as _intentionally_ closed. Maybe just remove them from the listeners map in `ListenClose`? That's probably the simplest way (and ensures that there's no funky race-condition while the listener is in the process of being closed). Then, in the defer, we can:

1. Close the listener if and only if it's still in the listeners map.
2. Log an error if and only if the listener was still in the map. (we clear the map on shutdown as well).

Reply:
I agree, but we have to close the listeners in `ListenClose` because if we just delete them from the map, they will never be closed. In the logic you describe, the listener is only closed in the defer if it is still present in the map, so if it was simply deleted before in `ListenClose`, it will never be closed.

Reply:
I'm saying that we _both_ close the listener _and_ remove it from the map in `ListenClose`.

Reply:
That way:

1. We can tell here if the listener was intentionally closed (is it still in the map or not).
2. ListenAddresses, and friends will return the correct answers immediately instead of waiting for this function to return asynchronously.

Reply:
That's what I did (at least I hope so) in the last commit. I use the presence of the listener in the map to conditionally close it in the defer, while the call to `ListenClose` closes it immediately.

Reply:
Uh, yes, yes you did.